src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -56,33 +61,301 @@ pub fn copy_to<G, F>(
-    let is_leader = usize::cast_from((sink_id, "leader").hashed()) % scope.peers() == worker_id;
+    let num_workers = scope.peers();
+    let leader_id = usize::cast_from((sink_id, "completion").hashed()) % num_workers;
+    let is_leader = worker_id == leader_id;
+
+    let mut builder = AsyncOperatorBuilder::new("CopyToS3-completion".to_string(), scope.clone());
+
+    let mut completion_input = builder.new_disconnected_input(&completion_stream, Pipeline);
 
-    let mut builder = AsyncOperatorBuilder::new("CopyToS3".to_string(), scope.clone());
+    builder.build(move |_| async move {
+        // fallible async block to use the `?` operator for convenience
+        let leader_cleanup_work = || async move {
+            info!(%sink_id, %worker_id, "s3 leader worker completion");
+            let sdk_config = aws_connection
+                .load_sdk_config(&connection_context, connection_id, InTask::Yes)
+                .await?;
+
+            let client = mz_aws_util::s3::new_client(&sdk_config);
+
+            // Remove the INCOMPLETE sentinel file to indicate that the upload is complete.
+            // This will race against other replicas who are completing the same uploads,
+            // such that the first replica to complete its uploads will delete the sentinel
+            // and the subsequent replicas shouldn't error if the object is already deleted.
+            // TODO: Should we also write a manifest of all the files uploaded?
+            mz_ore::task::spawn(|| "copytos3:completion", {
+                let client = client.clone();
+                let bucket = s3_key_manager.bucket.clone();
+                let incomplete_sentinel_key = s3_key_manager.incomplete_sentinel_key();
+                async move {
+                    info!(%sink_id, %worker_id, "removing INCOMPLETE sentinel file");
+                    client
+                        .delete_object()
+                        .bucket(bucket)
+                        .key(incomplete_sentinel_key)
+                        .send()
+                        .await?;
+                    Ok::<(), anyhow::Error>(())
+                }
+            })
+            .wait_and_assert_finished()
+            .await?;
+            Ok::<u64, anyhow::Error>(0)
+        };
+
+        let mut res = None;
+        while let Some(event) = completion_input.next().await {
+            match event {
+                AsyncEvent::Data(_ts, data) => {
+                    for evt in data {
+                        // We only expect to receive one event on this stream
+                        assert!(
+                            res.is_none(),
+                            "unexpectedly received more than 1 event on the completion stream!"
+                        );
+
+                        // If we're not the leader or if the completion event is an error,
+                        // return immediately to the response callback.
+                        if !is_leader || evt.is_err() {
```

Comment:
It's better for the check for `is_leader` to be the first first statement of this operator where it exits if it's not the one.

Also, I'd move the call of `onetime_callback` to a single place, when the frontier has advanced to the empty antichain

Reply:
each worker still needs to pass along its response to the `onetime_callback`, so we can't early exit before reading this message from the completion stream